File: search/pkg/dal/s3.go
package dal

import (
	"context"
	"io"

	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/shelepuginivan/carmen/search/pkg/config"
)

type S3 struct {
	client *s3.Client
	bucket string
}

func NewS3(cfg *config.Config) *S3 {
	client := s3.New(s3.Options{
		Region:       cfg.S3Region,
		Credentials:  credentials.NewStaticCredentialsProvider(cfg.S3AccessKey, cfg.S3SecretKey, ""),
		BaseEndpoint: &cfg.S3Endpoint,
		UsePathStyle: true,
	})

	return &S3{
		client: client,
		bucket: cfg.S3Bucket,
	}
}

func (ds *S3) GetDocument(ctx context.Context, key string) ([]byte, error) {
	res, err := ds.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: &ds.bucket,
		Key:    &key,
	})
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	return io.ReadAll(res.Body)
}

func (ds *S3) PutDocument(ctx context.Context, key string, body io.Reader) error {
	_, err := ds.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: &ds.bucket,
		Key:    &key,
		Body:   body,
	})
	return err
}

func (ds *S3) DeleteDocument(ctx context.Context, key string) error {
	_, err := ds.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: &ds.bucket,
		Key:    &key,
	})
	return err
}

File: extractor/src/s3.py
import boto3

from models import Config


class DocumentsBucket:
    def __init__(self, config: Config) -> None:
        self._bucket = config.s3_bucket
        self._client = boto3.client(
            "s3",
            endpoint_url=config.s3_endpoint,
            aws_access_key_id=config.s3_access_key,
            aws_secret_access_key=config.s3_secret_key,
            region_name=config.s3_region,
        )

    def get_object(self, key: str) -> bytes:
        res = self._client.get_object(Bucket=self._bucket, Key=key)
        return res["Body"].read()
